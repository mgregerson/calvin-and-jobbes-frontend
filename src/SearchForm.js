import React, { useState } from "react";
import "./SearchForm.css";

const placeholderForm = {};

/** SearchForm (Write docstring)
 *
 * Props:
 *       - initialFormData: {term: ""}
 *       - handleSearch: Func def passed down.
 * State:
 *       - FormData: {term: ""} (CompanyList) OR {name: ""} (JobList)
 *
 * CompanyList -> SearchForm
 * JobList -> SearchForm
 */

function SearchForm({ initialFormData = placeholderForm, handleSearch }) {
  const [formData, setFormData] = useState(initialFormData);

  /** Handles keystrokes in searchbar and updates formData */
  function handleChange(evt) {
    const fieldName = evt.target.name;
    const value = evt.target.value;

    setFormData((currData) => {
      currData[fieldName] = value;
      return { ...currData };
    });
  }
  /** Calls filter function passed in from parent as prop */
  function handleSubmit(evt) {
    evt.preventDefault();
    handleSearch(formData.term);
  }
  return (
    <div className="SearchForm mb-4 d-flex pt-4">
      <form onSubmit={handleSubmit}>
        <div className="row justify-content-center justify-content-lg-start gx-1">
          <div className="col-8">
            <input
              id="search-term"
              name="term"
              className="form-control form-control-lg"
              placeholder="Search"
              onChange={handleChange}
              value={formData.term}
              aria-label="Search"
            />
          </div>
          <div className="col-auto">
            <button className="btn search-btn btn-lg btn-primary">
              Search
            </button>
          </div>
        </div>
      </form>
    </div>
  );
}

export default SearchForm;
