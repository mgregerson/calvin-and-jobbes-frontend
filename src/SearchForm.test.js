import SearchForm from "./SearchForm";
import { render, fireEvent } from "@testing-library/react";

function testHandleSearch() {
  console.log("Called testHandleSearch");
}

const testInitialFormData = {
  term: "",
};

testHandleSearch = jest.fn();

describe("SearchForm Component", function () {
  it("renders without crashing", function () {
    render(
      <SearchForm
        initialFormData={testInitialFormData}
        handleSearch={testHandleSearch}
      />
    );
  });

  it("matches snapshot", function () {
    const { container } = render(
      <SearchForm
        initialFormData={testInitialFormData}
        handleSearch={testHandleSearch}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("Searches for a company or job", function () {
    const { container } = render(
      <SearchForm
        initialFormData={testInitialFormData}
        handleSearch={testHandleSearch}
      />
    );
    const searchInput = container.querySelector("#search-term");
    const submitBtn = container.querySelector(".search-btn");
    // Fill out the form
    fireEvent.change(searchInput, { target: { value: "and" } });
    fireEvent.click(submitBtn);
    // Search Submitted!
    expect(testHandleSearch).toHaveBeenCalled();
  });

  it("Dynamically updates the value of the search box with current input", function () {
    const { container } = render(
      <SearchForm
        initialFormData={testInitialFormData}
        handleSearch={testHandleSearch}
      />
    );
    const searchInput = container.querySelector("#search-term");
    const submitBtn = container.querySelector(".search-btn");
    // Fill out the form with one letter. Check that the value has changed
    fireEvent.change(searchInput, { target: { value: "a" } });
    expect(searchInput.value).toEqual("a");
    // Fill out with another letter. Check that it has changed again.
    fireEvent.change(searchInput, { target: { value: "an" } });
    expect(searchInput.value).toEqual("an");
    // Remove everything, check that it has changed.
    fireEvent.change(searchInput, { target: { value: "" } });
    expect(searchInput.value).toEqual("");
  });
});
