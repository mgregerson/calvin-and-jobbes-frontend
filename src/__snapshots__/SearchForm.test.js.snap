// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm Component matches snapshot 1`] = `
<div>
  <div
    class="SearchForm mb-4 d-flex pt-4"
  >
    <form>
      <div
        class="row justify-content-center justify-content-lg-start gx-1"
      >
        <div
          class="col-8"
        >
          <input
            aria-label="Search"
            class="form-control form-control-lg"
            id="search-term"
            name="term"
            placeholder="Search"
            value=""
          />
        </div>
        <div
          class="col-auto"
        >
          <button
            class="btn search-btn btn-lg btn-primary"
          >
            Search
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;
